### ISSUE TITLE
Load balancing: write endpoint to check with readiness probe

### ISSUE THREAD

**Issue by mfisher-idmod** [InstituteforDiseaseModeling/covasim#128]
@John-Sheppard to add more info and get developer assigned ASAP.

For load balancing, there is a dependency on the dev team to write an endpoint that can be checked with the probe. To do the load testing we did yesterday, John 'simulated' a readiness probe by leveraging the minReadySeconds of the Kubernetes deployment spec.

**Comment by John-Sheppard** [comment]
The startup cost of Covasim in the Azure / Kubernetes environment is quite high.  > 1 1/2 minutes on average.  While the application is starting, users receive an HTTP status of 504's ( Gateway timeout ) if their traffic is directed at that pod during startup.   Right now I am using Kubernetes minReadySeconds to direct the infrastructure to not send traffic the pods for a certain length of time.  This is a stop-gap measure.  We should use Kubernetes readiness and liveliness probes to monitor startup and readiness. 

To accommodate this, we need to create an API endpoint that Kubernetes can use to validate when the application is ready to service requests and that it is healthy.  We should use two different endpoints for the readiness and liveliness checks.  On the readiness check,  the server should return an HTTP status of 503 ( Service Unavailable ) when it is running through its initialization and a 200 after initialization.  For the healthiness check we should probably perform some calculation or resource check to determine if the application instance is in a healthy state on the infrastructure is running. 

These API's should be given their own endpoints, i.e. api/health/readiness and api/health/liveliness.

**Comment by cliffckerr** [comment]
@mfisher-idmod @John-Sheppard what's the issue with using one of the existing endpoints, e.g. `api/rpcs/get_version`, which is what the webapp uses?

**Comment by devclinton** [comment]
I agree with @cliffckerr . Since we don't have external dependencies like a database or shared cache like redis, we really just need to validate the API is up and working. I think the get version can be used for this purpose at the moment. If later we do add external dependency, we can make a true health check API.

**Comment by John-Sheppard** [comment]
If we can make the get_ver API check to make sure that application initialization has completed, then we could.  I don't think we should, I think we should be more intentional about an API instead of just grabbing one at get_version API.

**Comment by devclinton** [comment]
If get version reruns, the app has initialized. The api basically doesn't start returning until the full flask app has loaded

**Comment by devclinton** [comment]
Let me also take a look at nginx health checks for upstream services. It might be an option
https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/

**Comment by devclinton** [comment]
Let me look into nginx healthchecks. I think we can combine that with get_version

I am thinking something like
```
location /api {
    proxy_pass http://127.0.0.1:8097/;
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-For $remote_addr;
    health_check uri=rpcs/get_version;
  }
```

I  need to check how this behaves from the perspective of another load balancer though

**Comment by devclinton** [comment]
@cliffckerr @mfisher-idmod 
Worked with John. The Nginx approach wasn't viable as it is meant more for a load balancer. We went with a different approach

1. I ended up having to add a new API that returns static json. The rpcs was a POST which tends to be more fragile for healtchecks. This will be 4 additional lines to cova_app, so minimal impact
2. Added curl to docker image
3. Calling the gunicorn bind directly to avoid 504s from nginx

In addition, I pre-bytecode compiled the covasim library. This should improve startup times on clusters and makes sense. At moment, we were having to bytecode compile every run of a new pod which was was adding overhead to starts. I also added caching to the numba functions so after their first call, they will not need to be recompiled through llvm. 

I tested these changes locally and they appear to be working. @John-Sheppard is going to test tomorrow in the actual cluster. Once we have that test, I will open a PR
